src/ca/mcgill/cs/jetuml/diagram/ControlFlow.java

```diff
@@ -243,4 +246,124 @@ public boolean isInConstructorCall(Node pNode)
 		}
 		return false;	
 	}
+	
+	/*
+	 * Returns true if pNode is the ImplicitParameterNode that gets created in constructor call
+	 */
+	private boolean isImplicitParameterNodeInConstructorCall(Node pNode)
+	{
+		assert pNode != null;
+		return pNode.getClass() == ImplicitParameterNode.class && pNode.getChildren().size()>0 && isInConstructorCall(getFirstChild(pNode));
+	}
+	
+	private Node getFirstChild(Node pNode)
+	{
+		assert pNode.getChildren().size()>0;
+		return pNode.getChildren().get(0);
+	}
+	
+	private Optional<Edge> getConstructorEdge(Node pNode)
+	{
+		assert pNode != null && isInConstructorCall(pNode);
+		for(Edge edge : aDiagram.edges())
+		{
+			if ( pNode.getClass() == CallNode.class && edge.getEnd() == pNode && edge.getClass() == ConstructorEdge.class )
+			{
+				return Optional.of(edge);
+			}
+		}
+		return Optional.empty();	
+	}
+
+	/**
+	 * @param pEdge the Edge to get downstream elements from 
+	 * @return the downstream DiagramElements of pEdge
+	 * @pre pEdge!=null
+	 */
+	public Collection<DiagramElement> getEdgeDownStreams(Edge pEdge)
+	{
+		assert pEdge!=null;
+		Set<DiagramElement> downstreamElements = new HashSet<>();
+		// The edge addition here is necessary for recursive calls
+		downstreamElements.add(pEdge);
+		if(pEdge.getClass() == ConstructorEdge.class)
+		{
+			Node endParent = pEdge.getEnd().getParent();
+			downstreamElements.add(endParent);
+			downstreamElements.addAll(endParent.getChildren());
+			
+			// Recursively add downstream elements of the child nodes
+			for(Node child: endParent.getChildren())
+			{
+				for(Edge edge: getCalls(child))
+				{
+					downstreamElements.addAll(getEdgeDownStreams(edge));
+				}
+				// Add upstream edges of the child nodes
+				for(Edge edge: aDiagram.edges())
+				{
+					if(edge.getEnd() == child)
+					{
+						downstreamElements.add(edge);
+					}
+				}
+			}
+		}
+		else if(pEdge.getClass() == CallEdge.class)
+		{
+			CallNode endNode = (CallNode)pEdge.getEnd();
+			downstreamElements.add(endNode);
+			for(Edge e: getCalls(endNode))
+			{
+				downstreamElements.addAll(getEdgeDownStreams(e));
+			}
+		}
+		return downstreamElements;
+	}
+
+	/** 
+	 * @param pNode the Node to get downstream elements from 
+	 * @return the downstream DiagramElements of pNode
+	 * @pre pNode!=null
+	 */
+	public Collection<DiagramElement> getNodeDownStreams(Node pNode)
+	{

```

Comment:
Please rename to `getDownstreamNodes`

Reply:
The method would return all `DiagramElement`s in the downstream of the`pNode`, not necessarily only `Node`s, so I think the original name is more informative in this case?